### PR TITLE
Fix bf16 checkpoint round-trip and DDP val-loss reduction

### PR DESCRIPTION
**Pull request by workofart**
## Summary
- Persist each array's dtype in the checkpoint JSON sidecar and restore it on load. numpy's .npy format cannot store ml_dtypes extension dtypes like bfloat16, so bf16 checkpoints round-tripped as opaque `|V2` void bytes; the sidecar dtype lets the load path view-reinterpret the stored bytes back to the original dtype. Checkpoints saved before the dtype field existed still load as-stored.
- Save checkpoint arrays with `savez` instead of `savez_compressed`: dense weights barely compress and the deflate pass dominated checkpoint save time.
- Fix DDP validation-loss reporting: `metrics_row` only allreduced the val-loss pair inside the `log_global_loss` branch, so eval-only reporting (`log_global_loss=False`) returned each rank's local validation loss instead of the global one.

## Test plan
- New tests: dtype recorded in sidecar; bf16 round-trips through void npz bytes; checkpoints without dtype metadata still load; npz members are uncompressed; val loss is allreduced under DDP when train loss stays local (fails before the fix).
- Full suite: 557 passed, 11 skipped.

### CODE REVIEW

File: autograd/tools/trainer.py [hunk 1]
```diff
@@ -255,13 +255,13 @@ def _agg_dist_metrics(
...
                     ]
                 )
-                if eval_state is not None:
-                    values.extend(
-                        [
-                            xp.asarray(val_loss_sum, dtype=xp.float32),
-                            xp.asarray(val_loss_total_weight, dtype=xp.float32),
-                        ]
-                    )
+            if eval_state is not None:
```
Comment (workofart) [inline autograd/tools/trainer.py:258]: before, if we set `log_global_loss=False`, and then each rank evaluates its own val loss, but the val_loss_sum never gets all-reduced on line 268 because it's guarded by `if log_global_loss` on line 248, then the resulting `val_loss_sum` being reported will only contain the rank-local val loss, which is inaccurate.
